### PR TITLE
Stateful `IterableDataset`

### PR DESCRIPTION
**Pull request by ordabayevy**

### CODE REVIEW

File: cellarium/ml/core/datamodule.py [hunk 1]
```diff
@@ -226,3 +227,86 @@ def predict_dataloader(self) -> torch.utils.data.DataLoader:
...
+                raise ValueError(
+                    "Cannot resume training with a different drop_incomplete_batch value. "
+                    f"Expected {self.drop_incomplete_batch}, got {state_dict['drop_incomplete_batch']}."
+                )
+            if state_dict["n_train"] != self.n_train:
+                raise ValueError(
+                    "Cannot resume training with a different train size. "
+                    f"Expected {self.n_train}, got {state_dict['n_train']}."
+                )
+            if state_dict["worker_seed"] == self.worker_seed:
```
Comment (ordabayevy) [inline cellarium/ml/core/datamodule.py:306]: Don't throw a warning if it is `None`

File: cellarium/ml/data/dadc_dataset.py [hunk 2]
```diff
@@ -428,10 +440,19 @@ def __iter__(self):
         else:
             per_replica = math.ceil(n_obs / num_replicas)
         total_size = per_replica * num_replicas
+        batches_per_replica = len(self)
 
         # workers
         worker_id, num_workers = get_worker_info()
 
+        if self.resume_step is not None:
```
Comment (ordabayevy) [inline cellarium/ml/data/dadc_dataset.py:448]: Maybe change `None` to `0` when not resuming?

File: cellarium/ml/core/module.py [hunk 3]
```diff
@@ -383,3 +392,20 @@ def teardown(self, stage: str) -> None:
         if not self._cpu_transforms_in_module_pipeline:
             self.trainer.datamodule.collate_fn = self.trainer.datamodule.collate_fn.first_applied  # type: ignore[attr-defined]
             self._cpu_transforms_in_module_pipeline = True
+
+    def on_save_checkpoint(self, checkpoint: dict[str, Any]) -> None:
+        fit_loop = self.trainer.fit_loop
+        epoch_loop = fit_loop.epoch_loop
+        batch_progress = epoch_loop.batch_progress
+        if batch_progress.current.completed < batch_progress.current.processed:  # type: ignore[attr-defined]
```
Comment (ordabayevy) [inline cellarium/ml/core/module.py:400]: Handle the case when checkpointing happens at the end of the epoch.

File: cellarium/ml/core/module.py [hunk 4]
```diff
@@ -349,6 +349,15 @@ def on_train_epoch_end(self) -> None:
         if callable(on_train_epoch_end):
             on_train_epoch_end(self.trainer)
 
+        combined_loader = self.trainer.fit_loop._combined_loader
+        assert combined_loader is not None
+        dataloaders = combined_loader.flattened
+        for dataloader in dataloaders:
+            dataset = dataloader.dataset
+            set_epoch = getattr(dataset, "set_epoch", None)
+            if callable(set_epoch):
```
Comment (ordabayevy) [inline cellarium/ml/core/module.py:358]: Use `set_resume_step` callable

File: cellarium/ml/core/module.py (hunk 3 above)
Comment (ordabayevy) [reply cellarium/ml/core/module.py:400]: In this case `completed` is incremented by the PyTorch Lightning.

https://github.com/Lightning-AI/pytorch-lightning/blob/3627c5bfac704d44c0d055a2cdf6f3f9e3f9e8c1/src/lightning/pytorch/loops/fit_loop.py#L316-L317

File: cellarium/ml/core/datamodule.py [hunk 5]
```diff
@@ -226,3 +227,86 @@ def predict_dataloader(self) -> torch.utils.data.DataLoader:
...
+                raise ValueError(
+                    "Cannot resume training with a different number of nodes. "
+                    f"Expected {self.trainer.num_nodes}, got {state_dict['num_nodes']}."
+                )
+            if state_dict["batch_size"] != self.batch_size:
+                raise ValueError(
+                    "Cannot resume training with a different batch size. "
+                    f"Expected {self.batch_size}, got {state_dict['batch_size']}."
+                )
+            if state_dict["accumulate_grad_batches"] != 1:
```
Comment (mbabadi) [inline cellarium/ml/core/datamodule.py:279]: Let's make an issue for this.

**Review by mbabadi — APPROVED** [review]
Great work!

**Pull request merged**